src/main/java/org/dbos/apiary/postgres/PostgresContext.java

```diff
@@ -18,51 +20,54 @@
     // This connection ties to all prepared statements in one transaction.
     private final Connection conn;
     private long transactionId;  // This is the transaction ID of the main transaction. Postgres subtransaction IDs are invisible.
     private AtomicLong functionIDCounter = new AtomicLong(0);
     private long currentID = functionID;
 
-    public PostgresContext(Connection c, ProvenanceBuffer provBuff, String service, long execID, long functionID) {
-        super(provBuff, service, execID, functionID);
+    public PostgresContext(Connection c, WorkerContext workerContext, String service, long execID, long functionID) {
+        super(workerContext, service, execID, functionID);
         this.conn = c;
         this.transactionId = -1;
     }
 
     @Override
     public FunctionOutput apiaryCallFunction(String name, Object... inputs) {
-        Object clazz;
-        try {
-            clazz = Class.forName(name).getDeclaredConstructor().newInstance();
-        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
-            e.printStackTrace();
-            return null;
-        }
-        assert(clazz instanceof ApiaryFunction);
-        ApiaryFunction f = (ApiaryFunction) clazz;
-        // Remember current txid.
-        try {
-            Savepoint s = conn.setSavepoint();
-            long oldID = currentID;
+        ApiaryFunction f = workerContext.getFunction(name);
+        String functionType = workerContext.getFunctionType(name);
+        if (functionType.equals(ApiaryConfig.postgres) || functionType.equals(ApiaryConfig.stateless)) {
+            try {
+                Savepoint s = conn.setSavepoint();
+                long oldID = currentID;
+                try {
+                    this.currentID = functionID + functionIDCounter.incrementAndGet();
+                    FunctionOutput o = f.apiaryRunFunction(this, inputs);
+                    this.currentID = oldID;
+                    conn.releaseSavepoint(s);
+                    return o;
+                } catch (Exception e) {
+                    e.printStackTrace();
+                    this.currentID = oldID;
+                    conn.rollback(s);
+                    conn.releaseSavepoint(s);
+                    return null;
+                }
+            } catch (SQLException e) {
+                e.printStackTrace();
+                return null;
+            }
+        } else {
             try {
-                this.currentID = functionID + functionIDCounter.incrementAndGet();
-                FunctionOutput o = f.apiaryRunFunction(this, inputs);
-                this.currentID = oldID;
-                conn.releaseSavepoint(s);
-                return o;
+                ApiaryConnection c = workerContext.getConnection(functionType);
+                long newID = ((this.functionID + calledFunctionID.incrementAndGet()) << 4);
```

Comment:
Once we have the new cross-DB transaction implemented, this newID would be `this.currentID = functionID + functionIDCounter.incrementAndGet();` in order to be consistent with the synchronous invocation as above (line 46).

Reply:
It actually needs to be consistent with the IDs generated by queued functions in Apiary context.